Add BlobWriteOption to support MD5 and CRC32C checks on create/write

mziccard opened the pull request:
This PR fixes #267. Main changes are:
- Add `BlobWriteOption` with support for `md5Match` and `crc32cMatch`
- Add MD5 and CRC32C computation to `create(blobInfo, content, options)`
- Change `BlobTargetOption...` to `BlobWriteOption...` in create from stream writer
- Update unit tests
- Update and add integration tests

To note that `md5Match` and `crc32cMatch` are boolean options always true. If not specified storageobject's `crc32c` and `md5` fields are set to null (so no validation occurs).
Alternatives to boolean options were:
- String options implicitly taken from `BlobInfo`: this would have required to modify `storage.optionMap`, that is now used to fetch implicit values only for (meta)generation matches. (meta)generation options are shared across all APIs, so I felt like `md5Match` and `crc32cMatch` did not fit here.
- String options explicity set by the user (e.g. `BlobWriteOption.md5Match("7uyg8yg8==")`), still a valid option but a bit counter-intuitive: all values set in `BlobInfo` are transmitted, why not to use `md5` and `crc32c` fields?

Comments are appreciated!

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/spi/DefaultStorageRpc.java:
```diff
@@ -106,6 +108,15 @@ private static StorageException translate(GoogleJsonError exception) {
     return new StorageException(exception.getCode(), exception.getMessage(), retryable);
   }
 
+  private static void applyOptions(StorageObject storageObject, Map<Option, ?> options) {
```
I wonder if we can do this change without effecting the RPC layer.

So far all Options (defined in StorageRpc) are part of the apiary request headers (e.g. see [here](https://cloud.google.com/storage/docs/json_api/v1/objects/list) and [here](https://cloud.google.com/storage/docs/json_api/v1/objects/insert)).

See my suggestion bellow on how we can achieve that but if we decide to go ahead with the current way then at least lets make this method "less" inviting. "applyOptions" sounds like a generic thing that you might want to do but this should be done only by create and write.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java:
```diff
@@ -42,6 +42,8 @@
     IF_SOURCE_METAGENERATION_NOT_MATCH("ifSourceMetagenerationNotMatch"),
     IF_SOURCE_GENERATION_MATCH("ifSourceGenerationMatch"),
     IF_SOURCE_GENERATION_NOT_MATCH("ifSourceGenerationNotMatch"),
+    IF_MD5_MATCH("md5Hash"),
```
See my comment about avoid a change in the RPC layer but if we must then I would suggest naming them here as INCLUDE_MD5 and INCLUDE_CRC32C.

aozarov replied:
Actually, I am not sure about my naming suggestion now though I still hope we may not need them..

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -145,6 +145,67 @@ public static BlobTargetOption metagenerationMatch() {
+    }
+  }
+
+  class BlobWriteOption extends Option {
```
See my comment above about trying to isolate this change in the service impl layer.

Did you consider the following:
1) Change `BlobWriteOptions` to implement `Serializable` rather than extend `Option`.
2) Have a package scope static method in `BlobWriteOption` that accepts `BlobInfo` and `BlobWriteOption...`
and returns a tuple (`BlobInfo`, `BlobTargetOpiton...`).
3) Add another private `create` method similar to the one that accepts `InputStream` but this one accepts `BlobTargetOption`.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -524,7 +587,7 @@ public static Builder builder() {
    * @throws StorageException upon failure
    */
-  BlobInfo create(BlobInfo blobInfo, InputStream content, BlobTargetOption... options);
+  BlobInfo create(BlobInfo blobInfo, InputStream content, BlobWriteOption... options);
```
Shall we add a comment in the javadoc (here and in writer)  that "by default any md5 and crc32c values in the given blobInfo are ignored unless requested via the BlobWriteOption.md5Match and BlobWriteOption.crc32cMatch options"?

mziccard commented:
Fixed the comments. I removed options from the RPC layer. 
I don't like so much using `Tuple` for the PRC layer in the `BlobTargetOption.convert` method I added, what do you think?

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java:
```diff
@@ -123,20 +125,33 @@ public BucketInfo create(BucketInfo bucketInfo, BucketTargetOption... options) {
+    BlobInfo updatedInfo = blobInfo.toBuilder()
+        .md5(BaseEncoding.base64().encode(Hashing.md5().hashBytes(content).asBytes()))
+        .crc32c(BaseEncoding.base64().encode(
+            Ints.toByteArray(Hashing.crc32c().hashBytes(content).asInt())))
```
Is `Ints.toByteArray(Hashing.crc32c().hashBytes(content).asInt()))` != `Hashing.crc32c().hashBytes(content).asBytes` ?

mziccard replied:
The endianness changes. And according to [here](https://cloud.google.com/storage/docs/hashes-etags#_CRC32C):
> The Base64 encoded CRC32c is in big-endian byte order.

we need the big endian version.

aozarov replied:
Oh, right.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java:
```diff
@@ -123,20 +125,33 @@ public BucketInfo create(BucketInfo bucketInfo, BucketTargetOption... options) {
-      BlobTargetOption... options) {
-    final StorageObject blobPb = blobInfo.toPb();
-    final Map<StorageRpc.Option, ?> optionsMap = optionMap(blobInfo, options);
+  public BlobInfo create(BlobInfo blobInfo, final InputStream content, BlobWriteOption... options) {
```
Nit, `final` is not necessary here.

aozarov commented:
> I don't like so much using Tuple for the PRC layer in the BlobTargetOption.convert method I added, what do you think?

Honestly, I wished `Tuple` was part of the JDK (or at least guava) but I understand the opposition for it.
I am fine with it the way it and as this that is internal we can always be changed later.
If you feel better you can write a dedicate private inner class for it or pass a `BlobInfo.Builder`.

mziccard commented:
@aozarov I don't like the idea of doing side effects on a builder either, lets keep it as is, we can possibly change it later.

mziccard commented:
Removed the unnecessary `final`
